packages/react-api/src/Api.tsx

```diff
@@ -189,30 +191,44 @@ function Api ({ children, store, url }: Props): React.ReactElement<Props> | null
   );
 
   // initial initialization
-  useEffect((): void => {
-    const provider = new WsProvider(url);
-    const signer = new ApiSigner(queuePayload, queueSetTxStatus);
-    const types = getDevTypes();
-
-    api = new ApiPromise({ provider, registry, rpc: typesRpc, signer, types, typesBundle, typesChain, typesSpec });
-
-    api.on('connected', () => setIsApiConnected(true));
-    api.on('disconnected', () => setIsApiConnected(false));
-    api.on('error', (error: Error) => setApiError(error.message));
-    api.on('ready', (): void => {
-      const injectedPromise = web3Enable('polkadot-js/apps');
 
-      injectedPromise
-        .then(setExtensions)
-        .catch(console.error);
-
-      loadOnReady(api, injectedPromise, store, types)
-        .then(setState)
-        .catch((error): void => {
-          console.error(error);
-
-          setApiError((error as Error).message);
-        });
+  useEffect((): void => {
+    const create = (specName: SpecName, callback: () => void) => {
+      const provider = new WsProvider(url);
+      const signer = new ApiSigner(queuePayload, queueSetTxStatus);
+      const types = getDevTypes();
+
+      if (specName) {
+        api = new ApiPromise({ derives: derivesSpec[specName], provider, registry, rpc: typesRpc, signer, types, typesBundle, typesChain, typesSpec });
+      } else {
+        api = new ApiPromise({ provider, registry, rpc: typesRpc, signer, types, typesBundle, typesChain, typesSpec });
+      }
+
+      api.on('connected', () => setIsApiConnected(true));
+      api.on('disconnected', () => setIsApiConnected(false));
+      api.on('error', (error: Error) => setApiError(error.message));
+      api.on('ready', callback);
+    };
+
+    create(undefined, () => {
+      const specName = Object.keys(derivesSpec).find((specName) =>
+        api.runtimeVersion.specName.eq(specName)
+      ) as keyof typeof derivesSpec | undefined;
+
+      create(specName, () => {
```

Comment:
The problem with this is that the API init is a very heavy operation, so basically (as you'e indicated), we are hacking around the API. The correct approach would be to adjust `typesBundle` on the API to also take optional derives and let the API decorate.

Overall not unhappy with allowing the API to decorate derives based on info it has retrieved (this would be useful in a number of situations), just rather do it correctly.

Does it make sense?

Reply:
Thank you